ouroboros-network/src/Ouroboros/Network/PeerSelection/Governor.hs

```diff
@@ -504,15 +504,31 @@ peerSelectionGovernorLoop tracer
                           policy
                           jobPool
                           pst = do
-    loop pst
+    loop pst `catch` (\e -> traceWith tracer (TraceOutboundGovernorCriticalFailure e) >> throwIO e)
   where
     loop :: PeerSelectionState peeraddr peerconn
          -> m Void
     loop !st = assertPeerSelectionState st $ do
       -- Update public state using 'toPublicState' to compute available peers
       -- to share for peer sharing
       atomically $ writeTVar stateVar (toPublicState st)
+
       blockedAt <- getMonotonicTime
+
+      -- If by any chance the node takes more than 15 minutes to converge to a
+      -- clean state, we crash the node. This could happen in very rare
+      -- conditions such as a global network issue, DNS, or a bug in the code.
+      -- In any case crashing the node will force the node to be restarted,
+      -- starting in the correct state for it to make progress.
+      case bootstrapPeersTimeout st of
+        Nothing -> pure ()
+        Just t
+          | blockedAt >= t ->
+            error ("peerSelectionGovernorLoop: The node took more than 15 minutes "
+                  ++ "to converge to a clean state. Something is wrong!"
+                  )
```

Comment:
Let's add a proper exception and use `throwIO`.  Galois helped us to remove all unnecessary `error` calls, so let's be careful when we introduce new ones.  Since this is something we want to preserve, rather than a some internal `API` error / invariant vioilation it deserves an exception.

Reply:
Yes that makes sense